### ISSUE TITLE
Multiple encryption keys/key methods

### ISSUE THREAD

**Issue by sjau** [zfsonlinux/zfs#6824]
I'd like to propose a feature request to support multiple key slots/method slots like LUKS does it.

This can be very useful so that for example 2 different people have different passwords and they can independantly open up the encrypted dataset without needing to know the other's (highly secure and complex) passphrase.

Also, if you have multiple zpools it could be useful to provide path to the keyfile to automatically unlock it if needed...

**Comment by Lady-Galadriel** [comment]
I too desire multiple key slots. My thoughts are these:

1. Having a second key may allow recovery. Meaning you can have a master encryption key on a USB flash drive, kept locked up. But the normal operations staff have access to a another one with a password. If the password get's lost or changed without knowing what it is, the master can be used.

2. Additionally, an option to allow either R/W or R/O keys would be useful. This would affect mounting the dataset and ability to change the keys, (like operations for backups would only need R/O). But, there would be a safeguard to prevent deleting the last R/W key.

3. Also, perhaps a Master / Secondary flag for each key. This would allow 2 slots to both be R/W, but a Secondary could not delete / change the key or password from the Master slot. Again useful when dealing with operations staff who may perform scheduled changes involving reboots or pool imports. Plus, there should be a safeguard in place that would prevent the last Master key from being deleted.

For the last, it could very well be numerical order, and not Master / Secondary. That would be useful if we had more than 2 keys.

Having watched the chaos when users have used FreeNAS's Geli disk encryption, and not understood it, I would prefer a scheme that had more than 1 key slot. (Still won't protect users from themselves. But, this may protect Unix SysAdmins from having to perform restores when someone else looses or changes the key, and can't supply any key...)

**Comment by Lady-Galadriel** [comment]
@sjau, in regards to;

> Also, if you have multiple zpools it could be useful to provide path to the keyfile to automatically unlock it if needed...

This sort of works already. It's in the manual page, (which I don't have handy). I did submit a feature request to allow RAW keys to be stored on disk partitions, (like an USB flash drive), #6556

**Comment by sjau** [comment]
well the think with auto-unlock was how I used luks/dm-crypt... I can either provide the key file or enter manually the key... which is convenvient

**Comment by Lady-Galadriel** [comment]
All variations are supported if I remember correctly. Prompt or path to passphrase or key;

Prompt for passphrase
Prompt for location of key file

Path to file with passphrase
Path to file with key

As I said, I don't have the manual page handy to show examples.

**Comment by sjau** [comment]
the benefit is if you have multiple key slots... if you can provide a password or a path to file.. just whatever seems more convenient... e.g. have a keyfile in /root on your server, take out the disk, use it somewhere else and since the keyfile won't exist on the other system, provide password

**Comment by Lady-Galadriel** [comment]
Yes, that would be useful at times.

Plus. it would allow for split mirror pools. Basically have a mirrored pool, (at least 2 way, but more is supported), and use the `zpool split` command to create a second pool. One you could export and import elsewhere using a passphrase.

**Comment by sjau** [comment]
that's why I suggested it...

**Comment by sjau** [comment]
@tcaputi 

Is this possible to add as feature?

**Comment by tcaputi** [comment]
@sjau 
> Is this possible to add as feature?

Yes. The current implementation was actually designed with the ability to support multiple keys at some point. It has just not been a priority for me at the moment and I currently have a few other projects which I've committed to working on first. With regards to some of the specific features and use-cases that have been mentioned:

> This can be very useful so that for example 2 different people have different passwords and they can independantly open up the encrypted dataset without needing to know the other's (highly secure and complex) passphrase.

This is probably the biggest use case that I can think of.

> Having a second key may allow recovery. Meaning you can have a master encryption key on a USB flash drive, kept locked up. But the normal operations staff have access to a another one with a password. If the password get's lost or changed without knowing what it is, the master can be used.

This really just boils down to the above use case of having 2 people with different passphrases / keys. 

> Additionally, an option to allow either R/W or R/O keys would be useful. This would affect mounting the dataset and ability to change the keys, (like operations for backups would only need R/O). But, there would be a safeguard to prevent deleting the last R/W key.

> Also, perhaps a Master / Secondary flag for each key. This would allow 2 slots to both be R/W, but a Secondary could not delete / change the key or password from the Master slot. Again useful when dealing with operations staff who may perform scheduled changes involving reboots or pool imports. Plus, there should be a safeguard in place that would prevent the last Master key from being deleted.

Unfortunately, the cryptography here doesn't allow for this. All data is protected with symmetric key encryption, so there is no R/O key. Using public key cryptography would render the implementation uselessly slow. Without the ability to do this, it becomes impossible to have primary / secondary keys since then the secondary key could only be stopped from making changes via permissions which are trivial to patch over if you have root access to the machine. As a general rule, we have decided not to implement encryption functionality that cannot be enforced with cryptography so that we don't have this odd distinction between things that can / can't be done maliciously and things that can / can't be done inadvertently. My personal belief on the matter is that encryption and permissions need to be thought of differently and kept separate.

> Plus. it would allow for split mirror pools. Basically have a mirrored pool, (at least 2 way, but more is supported), and use the zpool split command to create a second pool. One you could export and import elsewhere using a passphrase.

Not quite sure if I understand here. Couldn't you just copy the key and use it for the new pool?

**Comment by sjau** [comment]
@tcaputi 

No worries... just take your time.. I know you're quite busy... but at least #8684 is merged now :)

I wasn't sure if you've seen this before, that's why I sent a message to you. It's a convenience feature and I'm glad to hear that the current encryption is designed with multi-slots use later on.

**Comment by tcaputi** [comment]
> Additionally, an option to allow either R/W or R/O keys would be useful. This would affect mounting the dataset and ability to change the keys, (like operations for backups would only need R/O). But, there would be a safeguard to prevent deleting the last R/W key.

As a quick side note, scrubs / resilvers can be accomplished without any loaded keys. Backups can be taken without keys as well if you use the `zfs send --raw` feature.

**Comment by sjau** [comment]
I think it's really great that scrubs / resilver / raw send don't required to load the keys. So even third parties can do that maintenance easily :)

**Comment by jknockaert** [comment]
Any chance this feature will be included in a 0.8 point upgrade?

**Comment by tcaputi** [comment]
In general, point releases are meant for bug fixes and so I would not expect this feature to show up in 0.8.x release. I am also not currently working on this feature at the moment.

**Comment by viggy96** [comment]
This feature would be awesome to have. I'm really paranoid about my keyfile being somehow corrupted or something, and my data being inaccessible without it. I will of course backup the keyfile to several locations, but still, one can't help but worry.

**Comment by tcaputi** [comment]
I agree that this feature would be nice to have, but I don't think it will help you with protecting your key file from corruption. At the end of the day keeping access to your data is just a matter of keeping access to a key that will work. Therefore, having a backup of the key is exactly the same as having a different key for this purpose. This feature is more about allowing multiple people to have their own passwords or keys to access the data.

**Comment by kithrup** [comment]
The general way to do this is to have an encrypted key, and then multiple keys which can be used to decrypt it.  (Tom knows this, I'm just not sure how much other people do, so sorry if it sounds condescending.)  So the change here would be to add user/key pairs somewhere, which would then be used to unlock the main key.

**Comment by viggy96** [comment]
@kithrup Isn't that the idea behind the current implementation? Aren't the user provided passwords/keys just used as a wrapper for the real encryption key?

**Comment by kithrup** [comment]
Yes, the user/key mapping just hasn't been done :).

**Comment by mddeff** [comment]
Consider another vote for interest in the feature.  I want to implement a clevis/tang unlocker for ZFS.  Use case is needing a breakglass key in case the tang server goes up in flames.  I like clevis/tang, but I'm not willing to run it (in production) without a backup way to decrypt the data.

**Comment by bill-mcgonigle** [comment]
Could we use properties for this securely, as an interim/0.8.x measure?
e.g.
```
# openssl enc -aes-256-cbc -pbkdf2 -a -A < tank-test.key > tank-test.key.aes
enter aes-256-cbc encryption password: ********
Verifying - enter aes-256-cbc encryption password: ********
# cat tank-test.key.aes
U2FsdGVkX19Ifd8hLvjm1RGFfA23zkNxE/GilHh9Z/32b6DDqeO44mkskgPcZy4ec1PQH/FowzPQLDZRKoP9OQ==#
# zfs set cryptkey:001='U2FsdGVkX19Ifd8hLvjm1RGFfA23zkNxE/GilHh9Z/32b6DDqeO44mkskgPcZy4ec1PQH/FowzPQLDZRKoP9OQ==' tank/test
# zfs get cryptkey:001 tank/test
NAME          PROPERTY      VALUE                                                                                     SOURCE
tank/test  cryptkey:001  U2FsdGVkX19Ifd8hLvjm1RGFfA23zkNxE/GilHh9Z/32b6DDqeO44mkskgPcZy4ec1PQH/FowzPQLDZRKoP9OQ==  local
```

And then running the inverse ( -d ) and setting keyformat=raw keylocation=outfile.   My interest is disaster recovery here; this would probably be inconvenient for frequent key swapping.  Some simple scripts could help.

I *think* this wouldn't work for pools with encrypted root datasets, but if somebody could correct me about the status of the root-level properties being encrypted or not that would be great.

**Comment by tcaputi** [comment]
@bill-mcgonigle yes that should work as long as you are happy with the sercurity of aes-cbc. I'm not actually all that familiar with the openssl command line tool, but this looks right. It should also work for pools with encrypted root filesystems. You can still import such pools to `zfs get` the properties you need. It just won't mount the root filesystem without the key.

**Comment by jknockaert** [comment]
@bill-mcgonigle I am not sure if I correctly understand what you're doing here. I assume that tank-test.key is the (unique) encryption key, and you're storing an encrypted version of it (encrypted using password `********`) as an arbitrary zfs property (which you call 'cryptkey:001' but it could effectively have any name not yet in use by an existing zfs property). So in fact you're using the zfs properties table to store an (externally) encrypted version of the zfs key. Am I correct?

**Comment by mdPlusPlus** [comment]
Is this currently being worked on? I was hoping I could replace LUKS with native ZFS encryption soon.

**Comment by robszy** [comment]
We need some form of recover password as it was said earlier but it is very important topic

**Comment by crocket** [comment]
Until /etc/init.d/zfs-load-key is merged and this is implemented, I use this workaround.

/usr/lib/dracut/modules.d/99local/module-setup.sh

```
#!/bin/sh

check() {
  return 0
}

depends() {
  echo zfs
  return 0
}

install() {
  inst_hook pre-mount 80 "${moddir}/zfs-load-key-all.sh"
}
```

/usr/lib/dracut/modules.d/99local/zfs-load-key-all.sh

```
modprobe zfs
zpool import -N -a
# Prevent linux kernel from printing messages over passphrase prompt
echo 3 > /proc/sys/kernel/printk

while true; do
  stty -echo
  read -p "Passphrase for datasets: " PASS
  stty echo
  echo "$PASS" | zfs load-key encryption-root1 || continue
  echo "$PASS" | zfs load-key encryption-root2 || continue
  break
done
```

**Comment by mdPlusPlus** [comment]
@crocket Isn't this something else? The issue is about having multiple keys for a single zpool/zfs dataset, not a single key for multiple zpools/zfs datasets.

**Comment by crocket** [comment]
Yes, it is something else. I just wanted one password to unlock everything.

**Comment by Djarid** [comment]
Is there any movement on this issue? It is the only thing stopping us from using zfs :(

**Comment by 1da1a172** [comment]
> Is there any movement on this issue? It is the only thing stopping us from using zfs :(

Until this gets implemented, you can use custom metadata fields to wrap the user key. I'm doing this to use the yubikey challenge/response mode to decrypt my root dataset.

**Comment by tobiasBora** [comment]
@1da1a172 Could you maybe elaborate on that solution for people not familiar with internals of ZFS?

And out of curiosity, why is it necessary to hardcode the number of slots? Is it only for simplicity to avoid to maintain a file whose size may change, or for another reason?

By the way, I was wondering: let's say that a malicious user has access to the harddrive with a valid key, even if later we change the key, ZFS would not prevent this malicious user to read the newly created data right? (I guess that all the user keys just decrypt a single master key which can't be changed without re-encoding the whole disk and could be learnt by a malicious user)… One way to avoid this would be to allow ZFS to define multiple master keys, and that when we change the key, a new master key is created to encode the newly written data… (old data don't need to be re-encoded anyway as the malicious adversary had already access to it). This way we don't need to re-encode the whole drive when a key is changed but then I guess it would be harder to program it as it means that the harddrive contains data encoded under multiple keys…)

**Comment by tcaputi** [comment]
@1da1a172 You don't need to use slots (if I did my job correctly, like 5 years ago) I planned for this to eventually be a feature.  See these #defines.

```
#define	DSL_CRYPTO_KEY_MASTER_KEY	"DSL_CRYPTO_MASTER_KEY_1"
#define	DSL_CRYPTO_KEY_HMAC_KEY		"DSL_CRYPTO_HMAC_KEY_1"
```

The idea was that you could just add more keys and give each a new number. I don't remember all of the details offhand, but I know I left some carveouts like this for supporting multiple keys

**Comment by 1da1a172** [comment]
@tobiasBora See `zfsprops(7)`, in particular, the "User Properties" section at the very end. The idea is that you would set a key the usual way, then encrypt it (through whatever means you want), and store the encrypted key in a user property. You could do this with multiple user properties and secrets. Then you have some scripting that takes the appropriate input to decrypt one of these encrypted keys, grabs the encrypted key from the user properties, does the decryption, then passes the key to ZFS.

For example, suppose you had `somezfskey` as a zfs user key for the dataset `pool/home`. You could run:
```bash
echo somezfskey | openssl enc -aes-256-ctr -pass pass:somepassphrase -pbkdf2 | base64
```
which outputs something like `U2FsdGVkX1/BNv5U8iczp0JLodFaPJ0hoZ5W`. This value can be stored in a user property:
```bash
zfs set pass1=U2FsdGVkX1/BNv5U8iczp0JLodFaPJ0hoZ5W pool/home
```
You would then load the key like so:
```bash
zfs-get -Ho value pass1 pool/home \
  | openssl enc -d -aes256-ctr -pbkdf2 -pass pass:somepassphrase \
  | zfs load-key -L prompt pool/home
```

**Comment by 1da1a172** [comment]
@tobiasBora Currently, the key that the user provides is used to encrypt/decrypt the master key, which is what actually encrypts/decrypts the data on disk. You can already change the user key, and in the new interface you will be able to add/remove user keys.

I don't think there will ever be multiple master keys. It complicates much and achieves little. If you were to add a new master key without reenctypting, only the new data would be protected by the new key, and the system would need both for the dataset to be usable.

Reencrypting (with a new master key) in place with a new master key is not in my scope of work.

**Comment by h0yle** [comment]
> ...you can use custom metadata fields to wrap the user key. I'm doing this to use the yubikey challenge/response mode to decrypt my root dataset...

Could you elaborate on this or provide a reference link? I would be interested to try this

> @1da1a172 You don't need to use slots (if I did my job correctly, like 5 years ago) I planned for this to eventually be a feature. See these #defines.
> 
> ```
> #define	DSL_CRYPTO_KEY_MASTER_KEY	"DSL_CRYPTO_MASTER_KEY_1"
> #define	DSL_CRYPTO_KEY_HMAC_KEY		"DSL_CRYPTO_HMAC_KEY_1"
> ```
> 
> The idea was that you could just add more keys and give each a new number. I don't remember all of the details offhand, but I know I left some carveouts like this for supporting multiple keys

Understanding that there are many competing priorities, having a backup key is important for disaster recovery and needs attention. Consider the possible case of a compromise where the attacker gains sufficiently elevated access to a system to change the loaded/running encryption key. The next time the backup system comes and pulls a snapshot will result in changing the encryption key of stored snapshots. A perfect recipe for ransom. This could also happen with fat fingers when updating a key, or a disgruntled employee. A well protected 2nd key stored in an offline vault would be very useful here.

**Comment by tcaputi** [comment]
I don't really have an updated status. I have not really been an active member of the zfs project since I switched jobs several years ago, to be honest. I would be happy to review any work towards this feature to the best of my ability, but I don't really have time to implement this between my new day job, my startup side job, and the extra consulting work I'm also doing. One of the maintainers may have more to say about what the current project priorities and timelines are.

That being said, I'm not sure if having a backup key built into the zfs filesystem would do a good job of preventing an attack such as the one you described. If an attacker has access to change the encryption key then they would almost definitely have access to manage (and destroy) any existing encrypted backup keys saved within zfs.

For any developers who might see this in the future, I think there is an answer here that wouldn't take too much work to implement. The current filesystem decryption code works using the user-supplied key material and the parameters saved on disk to decrypt the immutable master key. In the event of the attack described here, the stored parameters would no longer be available, but theoretically we could allow the user to save these parameters to a file that they could save securely on their own. Then they could decrypt the master key even if the on-disk encryption keys were altered via a new ioctl.

This isn't a very clean solution for allowing multiple users to access the dataset with different passwords, but it would probably be useful as a break-glass kind of solution. Perhaps @behlendorf might have thoughts on this? (Hi Brian, by the way. We haven't talked in a while :o) )

**Comment by amano-kenji** [comment]
Preparation for security breaches is not my concern.

I want multiple keys for booting a ZFS system with one password. With LUKS, it's possible to just unlock the root filesystem with one password, and the root filesystem has key files for other encrypted LUKS file systems.

With ZFS, I have to write a customized dracut hook. I'm sure 99.9% of linux users can't use a customized dracut hook. This is about user experience(UX).

If we don't make good UX, people will not use linux.

**Comment by robszy** [comment]
@tcaputi Bitlocker have option for backup keys to encrypted disks so in my opinion zfs also should have :)

**Comment by almereyda** [comment]
> theoretically we could allow the user to save these parameters to a file that they could save securely on their own. Then they could decrypt the master key even if the on-disk encryption keys were altered via a new ioctl.

For this use case, please also find similar conclusions at #12649.

**Comment by amano-kenji** [comment]
@f1d094 `zfs change-key -i` doesn't seem to do what I want unless a ZFS dataset can be unlocked by a password and also a key file. If I lose the key file, I want to unlock the dataset with a password.

**Comment by tcaputi** [comment]
> If we don't make good UX, people will not use linux. If you make it easy to use filesystem encryption, then more people will use it.

> Bitlocker have option for backup keys to encrypted disks so in my opinion zfs also should have :)

I understand the concern, but im just not really an active member of the project at the moment and I don't have the bandwidth to get back into it. I would be more than happy to talk to anyone / review any work someone makes towards a PR though.

**Comment by amano-kenji** [comment]
Is having ZFS on top of LUKS a bad idea?

**Comment by Majiir** [comment]
> Is having ZFS on top of LUKS a bad idea?

It works, but you lose encrypted incremental backups to untrusted backup hosts. That is IMO the primary motivation for ZFS native encryption in the first place.

Another way to get LUKS key management with ZFS is to create a tiny (unencrypted) zvol, use LUKS on that zvol, format the LUKS volume with something simple like ext4, and store the ZFS key there. I run this configuration on one of my machines to unlock a ZFS root in initrd using any of the LUKS keyslots.

**Comment by tcaputi** [comment]
I will add that there are some other advantages that come when using native zfs encryption as well. I am biased though, because this was my work. Off the top of my head:

- Per-data set encryption keys
- Hierarchal key management
- Ability to scrub the pool without encryption keys loaded
- Ability to use the pool at all without having encryption keys loaded
- MAC-based verification of data against malicious corruption
- Ability to send / receive datasets from untrusted sources / destinations

**Comment by Majiir** [comment]
> How does this protect you from the scenario I put forward?

The method I outlined does nothing for the scenario you described. This GitHub issue is titled "Multiple encryption keys/key methods" and the description asks for LUKS-like key management with ZFS.

> My core concern is the weaponizing of ZFS encryption against the host.

That is not what this issue is about, and there are issues linked earlier that would be better places to discuss your concern.

**Comment by tcaputi** [comment]
> Would this require any modification to the core components or could this be implemented as a standalone tool to include in zfsutils-linux, maybe?

The method I outlined would require a (small) kernel module change

> I decided to take the inititative and report this as security vulnerability:
> https://github.com/openzfs/zfs/security/advisories/GHSA-5wqj-fcr9-j434

The link appears to be broken.

In general, I think there are a lot of different use-cases / issues that all fall under the idea of having multiple keys to unlock a dataset. Again, I'm sorry I don't really have bandwidth to address these right now.

**Comment by Majiir** [comment]
@f1d094 https://github.com/openzfs/zfs/issues/12649 covers backup/restore of encryption keys.

**Comment by scineram** [comment]
>  Step 2: Malicious actor uses zfs change-key -i to re-encrypt zfs volumes
Just require the key to change it.

**Comment by sjau** [comment]
Wouldn't have thought that after 6 years, this suddenly gets so much attention again.

**Comment by almereyda** [comment]
There is a discussion on Practical ZFS about

* [The future of the ZFS native encryption - OpenZFS - Practical ZFS](https://discourse.practicalzfs.com/t/the-future-of-the-zfs-native-encryption/1037)

in which some developer has announced to be actively working on some of these issues. I have pointed them to the discussion here. If any organisation around is able to funnel funds towards their direction, it would be possible to use an existing momentum and build upon that.

Esp. now with the review offer.

**Comment by tcaputi** [comment]
@1da1a172 
Just to be sure we're all on the same page with the scope / current implementation (since there has been a lot of discussion). You are correct that the data is encrypted with the master key and the the master key is encrypted with the user key which is in turn derived from the user's password / key material along with some parameters stored on disk. All of these are not saved in a snapshot, and instead live in the top-level meta object store (MOS). When the user does a raw zfs send, the send file includes some extra instructions that REPLACE these on-disk values on the target system. So after a raw zfs recv has occurred, the old values are no longer accessible. Because of this, there isn't currently a good way to backup these values in a way where they will be permanently safe.

If we can save these parameters in a separate file which can be stored securely, these parameters should always be usable to decrypt the master key as long as the user remembers the corresponding password. This could be used in break-glass kind of situations, but it also comes with a (manageable) security risk that anyone who can access this file AND the password has a permanent way to access the dataset.

Theoretically, though, the implementation shouldn't be that hard. We just need to save the parameters needed to derive the user key to a file and then allow the proper hooks into the kernel to allow those parameters to be used instead of the ones that are saved in zfs. Then that would need to be wired out into the command line interface.

**Comment by amano-kenji** [comment]
@f1d094 It seems that you need to create a separate issue for that. That issue doesn't seem relevant to multiple key slots.

**Comment by shymega** [comment]
Agreed, it got a bit off-topic.

If multiple keyslots were implemented, I would state that we should put them behind a feature flag, and work with both Linux and (Free)BSD to bring their ZFS version/implementation up to date.

Has anyone from OpenZFS discussed multiple keyslots with the BSD and Linux folks?

**Comment by amano-kenji** [comment]
The problem is that native zfs encryption is unmaintained. Once somebody becomes dedicated to this, we can worry about BSD.

**Comment by amano-kenji** [comment]
It turns out that I needed only one key slot.

With keyformat=passphrase and keylocation=file:///path/to/keyfile, even if I lose the keyfile, I can reconstruct it if I remember the passphrase. I can just write the passphrase in keyfile.

I use the same passphrase for root filesystem and other filesystems. Before I knew this, I thought I needed one key slot for a keyfile and another key slot for a passphrase.

**Comment by anna328p** [comment]
+1 for this feature request. I want to be able to auto-unlock using a key stored in a TPM with a backup passphrase stored in hard-copy if the motherboard fails.

**Comment by amano-kenji** [comment]
Are you sure you want TPM? It will just unlock your computer without password. If an adversary has physical access to your computer, an unlocked computer can be read easily.

**Comment by anna328p** [comment]
I can use the TPM with Secure Boot attestation.

**Comment by Majiir** [comment]
@anna328p In case you missed the earlier comment, there is a [slightly convoluted workaround](https://github.com/openzfs/zfs/issues/6824#issuecomment-1818129528) you can use to combine LUKS key management with ZFS native encryption:

> Another way to get LUKS key management with ZFS is to create a tiny (unencrypted) zvol, use LUKS on that zvol, format the LUKS volume with something simple like ext4, and store the ZFS key there. I run this configuration on one of my machines to unlock a ZFS root in initrd using any of the LUKS keyslots.

As you described, I have a recovery key in one keyslot, and for the other I use a TPM and Secure Boot. I also use a PIN with the TPM. The effect is that I enter the PIN at boot, and my ZFS filesystems are unlocked and mounted. If I lose the PIN or the TPM PCRs don't match, I have to enter the recovery key.

Since the LUKS volume is a zvol on the same pool, it gets the same benefits from ZFS as other data (replication, bitrot protection, etc).

**Comment by amano-kenji** [comment]
Also, if your computer boots fully without any password, any attacker who stole your computer will not even need to circumvent encryption because the operating system is already fully functional without a password.

**Comment by amano-kenji** [comment]
@Majiir If you are worried about losing keyfile, you can use a passphrase keyfile which contains a passphrase. If you lose the keyfile, you can reconstruct it from your memory. You can remember the passphrase if you use it everyday.

**Comment by 1da1a172** [comment]
Well, I meant to post this a long time ago. It is another work-around using zfs user properties.
https://github.com/1da1a172/zfs-multisecret

**Comment by Majiir** [comment]
> @Majiir If you are worried about losing keyfile, you can use a passphrase keyfile which contains a passphrase. If you lose the keyfile, you can reconstruct it from your memory. You can remember the passphrase if you use it everyday.

I'm not worried about anything. I was just explaining the mechanics of the workaround.

> I unlock the root filesystem with one passphrase, and the other zfs file systems are unlocked with the same passphrase keyfile on the root filesystem.

I understand the scheme that you are using. However, it does not meet my needs, nor those of some others who have posted here. Multiple keyslots are necessary in order to have a recovery key in addition to a normal passphrase that's tied to a security key or TPM, for example.

**Comment by 1da1a172** [comment]
The solution I posted gives you multiple key slots. It just does it by layering on top of what already exists instead of adding it as a first party feature (which would obviously be better).

**Comment by anna328p** [comment]
@amano-kenji I'm not worried about state-level adversaries backdooring my TPM. I'm not worried about being forced to sign stuff with Microsoft's keys (my laptop, like all x86_64 machines with secure boot tend to, allows custom key management) or MS somehow spying on me via some data I load into a chip on my motherboard. And I require a login password on my machine.

My use-case is convenience and allowing unattended boot while making it difficult to read personal information from my laptop if it is stolen or lost.

@Majiir Thank you for that workaround, it's really clever and I will probably do something like that on my next install.

**Comment by amano-kenji** [comment]
@anna328p Unattended boot means any hacker who has physical access to your computer can read your computer.

At the very least, you should protect TPM with a password.

**Comment by anna328p** [comment]
> @anna328p Unattended boot means any hacker who has physical access to your computer can read your computer.

...if they have arbitrary code execution inside getty or gdm.

No, they can't just freeze and pull the RAM, my CPU has transparent memory encryption and the RAM on my laptop is soldered.

Either way, my threat model isn't a hacker who would have detailed knowledge of exploiting obscure unix setups, it's someone who steals my computer or picks it up if I lose it.

**Comment by amano-kenji** [comment]
If NSA steals your laptop, they will boot it up and somehow read it. TPM has already been backdoored for state-level adversaries. But, your threat model isn't NSA or mossad.

**Comment by norpol** [comment]
@amano-kenji can you please stop this? This has nothing to do with the ticket and also is completely out of place to unsolicitely educate someone on your personal opinions on TPM and state level attackers.

**Comment by lockheed** [comment]
Just adding myself here to follow this issue. I think it is critical and I was quite disappointed when I found out I still cannot replace LUKS with native ZFS encryption, even though the ticket is almost 8 years old.

**Comment by alkisg** [comment]
I too would like to see support for multiple keys.

For example, if an employee goes AWOL, a sysadmin with a second key can unlock their ZFS dataset, and the company can assign the documents to another employee.

**Comment by vuongtuha** [comment]
Sry i just jumped from gg search, but this's still not consider as feature, in 2025, right?
The most close thing i only can create LUKS on key partition, then unlock LUKS gave me these 2 opt, and this partition contain key for zfs auto load-key into pool

**Comment by robn** [comment]
It's still a reasonable thing to want, which is why this feature request is still open. It just doesn't seem to be a high priority for anyone with the interest and knowledge to implement it.

(fwiw, I'd rate it as a "medium" difficulty. It doesn't really require any crypto knowledge, though obviously some awareness and respect for the nature of the problem helps. Mostly, it's extending structures to store and manage N wrapping keys instead of just one, and plumbing it back up the stack to the CLI tools).
